src/combinatorics/additive/energy.lean

```diff
@@ -0,0 +1,66 @@
+/-
+Copyright (c) 2022 Yaël Dillies Ella Yu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies Ella Yu
+-/
+import data.finset.prod
+
+/-!
+# Additive energy
+
+This file defines the additive energy of two finsets of a group. This is a central quantity in
+additive combinatorics.
+-/
+
+variables {α : Type*} [decidable_eq α]
+
+namespace finset
+section has_mul
+variables [has_mul α] {s s₁ s₂ t t₁ t₂ : finset α}
+
+/-- The multiplicative energy of two finsets `s` and `t` in a group is the number of quadruples
+`(a₁, b₁, a₂, b₂) ∈ s × t × s × t` such that `a₁ * b₁ = a₂ * b₂`. -/
+@[to_additive additive_energy "The additive energy of two finsets `s` and `t` in a group is the
+number of quadruples
+`(a₁, b₁, a₂, b₂) ∈ s × t × s × t` such that `a₁ + b₁ = a₂ + b₂`."]
+def multiplicative_energy (s t : finset α) : ℕ :=
+(((s ×ˢ s) ×ˢ t ×ˢ t).filter $ λ x : (α × α) × α × α, x.1.1 * x.2.1 = x.1.2 * x.2.2).card
```

Comment:
I think it might be nice to have the finset itself as a definition too because that's sometimes useful, but I'm happy for that to be TODO for now.

Reply:
Added as a todo.